### ISSUE TITLE
sendFile not working

### ISSUE THREAD

**Issue by cherkavi** [cdapio/netty-http#68]
DefaultHttpHeaders headers = new DefaultHttpHeaders();
        headers
                .set("Content-type", "application/yaml")
                .set("Content-Disposition", "attachment; filename=\"sen.yaml\"")
        ;
        try {
            // responder.sendFile(tempFile.toFile(), headers);
            responder.sendFile(tempFile.toFile());
        } catch (IOException e) {
            responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "can't send file: "+e.getMessage());
        }

**Comment by chtyim** [comment]
Can you give more details? What is the error / behavior you see?

**Comment by robert-engel** [comment]
Quite an old issue but I also encountered a problem with `sendFile`. There were no errors in the log but chrome would display the error `ERR_CONTENT_DECODING_FAILED`.

I looked at the official netty example for http file transfer and saw that it was pretty much the same code as in `BasicHttpResponder#sendFile`. But running the netty example I had no issue serving the same file `sendFile` was failing with.

The netty example didn't have the `HttpContentCompressor` so I removed the compressor from the pipeline and then `sendFile` worked. Apparently there is an issue with combining `HttpContentCompressor` and `ChunkedWriteHandler`. I am not quite sure what the best way to fix the issue would be though.

Can this issue be avoided somehow or is `sendFile` currently just not working?

**Comment by chtyim** [comment]
The `sendFile` method is using `FileRegion` underneath for zero-copy writing to the channel. This means it won't work with the content compressor, otherwise it is not zero copy. However, when SSL is enabled, then the `sendFile` is using `HttpChunkInput` instead, which is compatible with the compressor.

This is indeed a problem in the library, and the fix could be bypassing the `HttpContentCompressor` when the `sendFile` method is being call to avoid it sending a wrong `Content-Encoding` header to the client.

**Comment by robert-engel** [comment]
The way netty checks for multiplicity is that a flag gets set on the compressor when its added the first time and even though its the same compressor you remove and then try to re add the check for multiplicity fails and the compressor will not be re attached I think.

```
io.netty.channel.ChannelPipelineException: io.netty.handler.codec.http.HttpContentCompressor is not a @Sharable handler, so can't be added or removed multiple times.
	at io.netty.channel.DefaultChannelPipeline.checkMultiplicity(DefaultChannelPipeline.java:597) ~[netty-transport-4.1.16.Final.jar:4.1.16.Final]
	at io.netty.channel.DefaultChannelPipeline.addAfter(DefaultChannelPipeline.java:318) ~[netty-transport-4.1.16.Final.jar:4.1.16.Final]
	at io.netty.channel.DefaultChannelPipeline.addAfter(DefaultChannelPipeline.java:308) ~[netty-transport-4.1.16.Final.jar:4.1.16.Final]
	at io.cdap.http.internal.BasicHttpResponder$4.run(BasicHttpResponder.java:265) ~[netty-http-1.4.0-SNAPSHOT.jar:?]
	at io.cdap.http.internal.BasicHttpResponder$1.operationComplete(BasicHttpResponder.java:141) ~[netty-http-1.4.0-SNAPSHOT.jar:?]
	at io.cdap.http.internal.BasicHttpResponder$1.operationComplete(BasicHttpResponder.java:138) ~[netty-http-1.4.0-SNAPSHOT.jar:?]
```

**Comment by chtyim** [comment]
Fixed by #81

**Comment by chtyim** [comment]
Good catch. Didn't notice there is a subtle log in the test output
